transcom/mymove
# MB-299 Migrate orders to using entities

## Description

Migrates redux pattern to using entities for **orders** and **uploads** for the PPM service member signup flow. 
This ticket was originally intended to only migrate the createOrders API func to entities, but as I started the work, I found that related objects needed to be loaded onto entities, i.e. documents and uploads. So, I added the modules funcs to documents to create and delete uploads.

Note:
* There is now a separate Uploader, called "OrdersUploader" to use with the swaggerRequest call & payload for adding and deleting uploads to integrate with filepond. This could potentially be combined back into the Uploader, since most of the code is repeated, but it may be more desirable to have a clearer separation between the components that populate entities with normalized data vs. not (i.e. the "old" model).
* The API endpoints that fetch orders for a logged in user are erroring in master due to a couple most-likely related bugs. These should be fixed with this PR for the SM flow purposes, but since these API changes are used elsewhere they should be tested with other apps i.e. Office, TOO, etc.. (see https://dp3.atlassian.net/browse/MB-2951 and https://dp3.atlassian.net/browse/MB-3007)
* The swaggerRequest has been adapted to allow for deleting arrays from the state store, in addition to just removing a single entity.

## Setup

Add any steps or code to run in this section to help others prepare to run your code:

```sh
# populate with test users
make db_dev_reset (if necessary)
make db_dev_migrate (if necessary)
bin/generate-test-data --named-scenario="e2e_basic"

make server_run
make client_run
```

Since this affects the entire SM and PPM sign up flow, the following cases should be tested:

Note: You should expect to see the redux state populate orders, uploads, documents, and moves @ `entities` rather than in their respective, current areas. I.e. when uploads are existent, they should appear @ `entities.uploads` rather than `orders.currentOrders.uploads`. (Note that `orders.CurrentOrders` is continuing to be populated via the getLoggedInUser endpoint as well, but only used and referenced in the entities.)

1. Sign up as a new service member in the local sign in flow and go through to the Review page. Click on Edit Orders and interact with the uploader and Uploadtable. 
2. Sign in as a service member and enter the PPM flow at various points. Log out, log back in, refresh, etc.
3. Sign in as a PPM office user, find the move and cancel it. Sign back in as a Service Member - You should see "start new move" or "your move has been canceled"

## Code Review Verification Steps

 [Querying the Database Safely](https://github.com/transcom/mymove/wiki/backend#querying-the-database-safely)
 have been satisfied.
* [ x] Request review from a member of a different team.
* [x ] Have the Jira acceptance criteria been met for this change?

## References
Jira Stories:
* https://dp3.atlassian.net/browse/MB-298
* https://dp3.atlassian.net/browse/MB-299
* https://dp3.atlassian.net/browse/MB-300
* https://dp3.atlassian.net/browse/MB-3007
* https://dp3.atlassian.net/browse/MB-2951
